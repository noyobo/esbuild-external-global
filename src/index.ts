import {BuildOptions} from "esbuild";
import path from "node:path";
import fs from 'fs-extra'


export const createExternalAlias = (alias: Record<string, string>) => {
  const aliasKeys = Object.keys(alias)

  const aliasMap = {}

  aliasKeys.forEach((key) => {
    const file = path.join(__dirname, 'alias', alias[key] + '.js')
    fs.ensureFileSync(file)
    fs.writeFileSync(file, `module.exports = ${alias[key]}`)
    aliasMap[key] = file
  })

  return aliasMap as BuildOptions['alias']
}

export const externalGlobal = (alias: Record<string, string>): (options: BuildOptions) => BuildOptions => {
  const aliasMap = createExternalAlias(alias)

  return (options: BuildOptions) => {
    const alias = options.alias || {}
    for (const key in aliasMap) {
      if (alias[key]) {
        throw new Error(`alias ${key} already exists`)
      }
      alias[key] = aliasMap[key]
    }
    return {
      ...options, alias
    }
  }
}
