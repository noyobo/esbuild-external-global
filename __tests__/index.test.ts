import esbuild from 'esbuild';
import {describe, expect, it} from "vitest"
import path from 'path';

import {externalGlobal} from '../src';


describe('external-global', () => {
  it('should bundle', async () => {
    const result = await esbuild.build(
      externalGlobal({
        react: 'window.React',
      })({
        entryPoints: [path.resolve(__dirname, 'fixtures/react/index.tsx')],
        bundle: true,
        write: false,
      }));

    expect(result.warnings).toHaveLength(0);
    expect(result.outputFiles).toHaveLength(1);
    expect(result.outputFiles[0].text).toMatchSnapshot()
  });
});
