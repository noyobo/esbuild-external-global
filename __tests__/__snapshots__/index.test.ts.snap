// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`external-global > should bundle 1`] = `
"(() => {
  var __create = Object.create;
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __getProtoOf = Object.getPrototypeOf;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __commonJS = (cb, mod) => function __require() {
    return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
    // If the importer is in node compatibility mode or this is not an ESM
    // file that has been converted to a CommonJS file using a Babel-
    // compatible transform (i.e. "__esModule" has not been set), then set
    // "default" to the CommonJS "module.exports" for node compatibility.
    isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
    mod
  ));

  // src/alias/window.React.js
  var require_window_React = __commonJS({
    "src/alias/window.React.js"(exports, module) {
      module.exports = window.React;
    }
  });

  // __tests__/fixtures/react/index.tsx
  var import_react = __toESM(require_window_React());
  var App = () => {
    const [count, setCount] = import_react.default.useState(0);
    return /* @__PURE__ */ import_react.default.createElement("div", { onClick: () => {
      setCount((prev) => prev + 1);
    } }, "App ", count);
  };
})();
"
`;
